lib/backends/webgl/webgl-context.ts

```diff
@@ -464,4 +468,64 @@ export class WebGLContext {
       }
     }
   }
+
+  beginTimer(): WebGLQuery {
+    if (this.version === 2) {
+      const gl2 = this.gl as WebGL2RenderingContext;
+      const ext = this.disjointTimerQueryWebgl2Extension;
+
+      const query = gl2.createQuery() as WebGLQuery;
+      gl2.beginQuery(ext.TIME_ELAPSED_EXT, query);
+      return query;
+    } else {
+      // TODO: add webgl 1 handling.
+      throw new Error(`Not support WebGL 2`);
+    }
+  }
+
+  endTimer() {
+    if (this.version === 2) {
+      const gl2 = this.gl as WebGL2RenderingContext;
+      const ext = this.disjointTimerQueryWebgl2Extension;
+      gl2.endQuery(ext.TIME_ELAPSED_EXT);
+      return;
+    } else {
+      // TODO: add webgl 1 handling.
+      throw new Error(`Not support WebGL 2`);
+    }
+  }
+
+  isTimerResultAvailable(query: WebGLQuery): boolean {
+    let available = false, disjoint = false;
+    if (this.version === 2) {
+      const gl2 = this.gl as WebGL2RenderingContext;
+      const ext = this.disjointTimerQueryWebgl2Extension;
+
+      available = gl2.getQueryParameter(query, gl2.QUERY_RESULT_AVAILABLE);
+      disjoint = gl2.getParameter(ext.GPU_DISJOINT_EXT);
```

Comment:
Curious in what scenario would available be false?

Reply:
The result  is not available until the profiled kernel execution completes.